### ISSUE TITLE
During rolling update in Kubernetes data in KV store is lost in some nodes

### ISSUE THREAD

**Issue by yoadey** [nats-io/nats-server#4351]
We are having a problem with our NATS cluster in a Kubernetes cluster, were the entries in KV stores differ betwenn the nats nodes after server restarts, leaving us with data loss in some cases. As the Kubernetes cluster gets updated every week, this issue was now seen several times, although we cannot reproduce it in a stable manner.
Our current workaround is, that if we don't get the data in the first run, we do the same request up to 5 times and then we normally get the wanted data due to the server request routing.

## Defect

 - [x] Included `nats-server -DV` output
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats-server` and affected client libraries used:

```
[68] 2023/07/31 10:13:00.901015 [INF] Starting nats-server
[68] 2023/07/31 10:13:00.901069 [INF]   Version:  2.9.20
[68] 2023/07/31 10:13:00.901073 [INF]   Git:      [not set]
[68] 2023/07/31 10:13:00.901079 [DBG]   Go build: go1.20.6
[68] 2023/07/31 10:13:00.901091 [INF]   Name:     ND2HTZ5KWVBYGEJPIJBGHBVKSPXTUSWULKAOKXKSFLQNW23C6RYO7DTC
[68] 2023/07/31 10:13:00.901100 [INF]   ID:       ND2HTZ5KWVBYGEJPIJBGHBVKSPXTUSWULKAOKXKSFLQNW23C6RYO7DTC
[68] 2023/07/31 10:13:00.901175 [DBG] Created system account: "$SYS"
```
(We used a custom build from 2.9.20, where we included this specific fix: https://github.com/nats-io/nats-server/issues/4320, Problem also occured with offical 2.9.19 nats-server container)

All client libraries are affected.

#### OS/Container environment:

Kubernetes containerd

#### Steps or code to reproduce the issue:

Have the following cluster setup:
- 3 NATS nodes
- Some (~1000) KV stores with replicas set to 3 and persistence set to file
- Write random entries to the KV stores (and keep writing/updating)
- restart one node one by one (rolling update of cluster/nats)
- check all entries exist in all KV stores on all nats servers/nodes, e.g. with requesting multiple times all entries in the same KV Store with the request `nats kv history <keyvaluestore> --trace` (trace option shows the server)

#### Expected result:

Requesting multiple times all entries in the same kv store should always return the same list.

#### Actual result:

Sometimes after restarts the result differs between the nodes and stay different.

#### Additional details

This does not occur after every restart, seem also to depend a little bit on the downtime from the node and for some KVs it seem to heal itself after some time (sometimes minutes, sometime hours, sometime days).

#### Go code for testscript:

Following script executes listing the entries in the KV store multiple times against the KV store and prints the result.
As the NATS server routes the traffic by itself, we don't need to switch servers.

```golang
package main

import (
	"log"
	"os"
	"strings"

	"github.com/nats-io/nats.go"
)

func connect() nats.JetStreamContext {
	url := os.Getenv("NATS_URL")
	if url == "" {
		url = nats.DefaultURL
	}

	nc, err := nats.Connect(url)
	if err != nil {
		log.Fatal(err)
	}
	js, err := nc.JetStream()
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("connected %s", url)
	return js
}

func kvNames() []string {

	js := connect()

	list := make([]string, 0, 1024)
	for r := range js.KeyValueStoreNames() {
		list = append(list, r[3:])
	}
	return list
}

func main() {
	log.Print("natscheck ...")

	list := kvNames()

	js := connect()
	for i, kvname := range list {
		log.Printf("%d/%d check buckets:%s", i, len(list), kvname)
		var lastList []string
		var entries int = -1
		for loop := 0; loop < 10; loop++ {

			kv, err := js.KeyValue(kvname)
			if err != nil {
				log.Fatalf("[%s]:%v", kvname, err)
			}
			currList, err := kv.Keys()
			if err != nil && !strings.Contains(err.Error(), "no keys found") {
				log.Fatal(err)
			}
			//kv.History()
			if loop > 1 {
				diff1 := difference(currList, lastList)
				if len(diff1) > 0 {
					log.Printf("Found %d/%d different keys in current, but not last list: %s", len(diff1), len(currList), strings.Join(diff1, ", "))
					entries = -1
				}
				diff2 := difference(lastList, currList)
				if len(diff2) > 0 {
					log.Printf("Found %d/%d different keys in last, but not current list: %s", len(diff2), len(lastList), strings.Join(diff2, ", "))
					entries = -1
				}
			} else {
				entries = len(currList)
			}
			lastList = currList
		}
		if entries > 0 {
			log.Printf("Found %d same entries", entries)
		}
	}
}

func difference(a, b []string) []string {
	mb := make(map[string]struct{}, len(b))
	for _, x := range b {
		mb[x] = struct{}{}
	}
	var diff []string
	for _, x := range a {
		if _, found := mb[x]; !found {
			diff = append(diff, x)
		}
	}
	return diff
}
```

**Comment by derekcollison** [comment]
When doing restarts, do you wait for /healthz to return ok before moving to next server?

**Comment by derekcollison** [comment]
Also be good to make sure we can reproduce from top of main, the 2.9.21 RC candidate.

**Comment by yoadey** [comment]
The readiness check waits for /healthz to return ok, and before that we don't server any traffic to that nats server.
However, I have the impression, that the new nats server registers itself at the other nats severs before its /healthz returns ok and those ready nats nodes serve requests to it. But that's just an impression from the cpu and memory stats and some logs I had, which I can't say reliably.

**Comment by derekcollison** [comment]
For the direct GETS, which all stream peers can participate in yes once they think they are caught up they can participate..

**Comment by wallyqs** [comment]
One thing that you could try too is to bump the `initialDelaySeconds` to be longer during your rollouts:
```
    liveness:
      enabled: true
      initialDelaySeconds: 100
```
https://github.com/nats-io/k8s/blob/f87652b54cbcbee6c32f7a3392982f4f690f0dc2/helm/charts/nats/values.yaml#L51C7-L51C26

**Comment by yoadey** [comment]
Will try the change in the initialDelaySeconds for the next restart next week, we currently want to restart as few times as possible as only the production system hast currently the load where we can monitor this with good data. We are trying to setup another test instance where we can reproduce this behavior, but until we have it finished we have to wait.

**Comment by yoadey** [comment]
Just tested the `liveness.initialDelaySeconds` set to 120 but it didn't helped, also for `readiness.initialDelaySeconds`.
One thing to note: the state before the test was good, meaning we didn't had any issues as the nats cluster seemed to have healed itself this morning when we made a rolling update to the application which uses the nats.

**Comment by derekcollison** [comment]
Keep us posted, you should not be losing any messages of course.

**Comment by wallyqs** [comment]
@mpwalkerdine were you using interest streams?

**Comment by mpwalkerdine** [comment]
Hi @wallyqs - no, permanent streams with durable consumers. A couple of potentially interesting things appeared in our logs around this time too, `Resetting WAL state` and `Stream state too short (0 bytes)` :thinking:

**Comment by mpwalkerdine** [comment]
A quick update for any future readers - we've narrowed the issue down somewhat and it appears to be related to the NATS disks being pulled at inappropriate times by the CSI plugin, leading to corruption and the log line during next startup:  
 `Detected orphaned stream '[ACC] > [NAME]' will cleanup`, which ultimately causes that node to delete it's message storage on startup.

**Comment by mpwalkerdine** [comment]
@wallyqs FYI it looks like one or more of the fixes in 2.10.10 has resolved our issue - I can no longer reproduce the deletion on startup issue, no orphaned stream logs and no propagation of that empty state to other nodes in the cluster :crossed_fingers:

**Comment by yoadey** [comment]
For me, the issue seems to be solved with the latest version. We are not able to reproduce it anymore, so it could be closed

**Comment by Jarema** [comment]
@mpwalkerdine Could you let us know if 2.10.11 fixed it for you too?

**Comment by mpwalkerdine** [comment]
The behaviour has changed, although there still seems to be a discrepancy between nodes when they come back up:

Health node:

```
State:

             Messages: 789
                Bytes: 5.0 MiB
             FirstSeq: 1 @ 2024-02-21T13:29:29 UTC
              LastSeq: 789 @ 2024-02-21T13:43:32 UTC
     Active Consumers: 321
   Number of Subjects: 95
```

Faulty node:

```
State:

             Messages: 543
                Bytes: 2.4 MiB
             FirstSeq: 247 @ 2024-02-21T13:42:58 UTC
              LastSeq: 789 @ 2024-02-21T13:43:32 UTC
     Active Consumers: 321
   Number of Subjects: 3
```

**Comment by wallyqs** [comment]
Fixed via #5821, thanks a lot @yoadey for the reproduction at https://github.com/yoadey/nats-test that was very useful.
